h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/exec/LocalXGBoostExecutor.java

```diff
@@ -0,0 +1,165 @@
+        modelKey = key;
+        rt = new RabitTrackerH2O(init.num_nodes);
+        BoosterParms boosterParams = BoosterParms.fromMap(init.parms);
+        boolean[] nodes = new boolean[H2O.CLOUD.size()];
+        for (int i = 0; i < init.num_nodes; i++) nodes[i] = init.nodes[i] != null;
+        MatrixLoader loader = new RemoteMatrixLoader(init.matrix_dir_path, init.nodes);
+        setupTask = new XGBoostSetupTask(
+            modelKey, null, boosterParams, init.checkpoint_bytes, getRabitEnv(), nodes, loader
+        );
+        featureMapFile = createFeatureMapFile(init.featureMap);
+    }
+
+    /**
+     * Used when executing from a local model
+     */
+    public LocalXGBoostExecutor(XGBoostModel model, Frame train) {
+        modelKey = model._key;
+        XGBoostSetupTask.FrameNodes trainFrameNodes = XGBoostSetupTask.findFrameNodes(train);
+        rt = new RabitTrackerH2O(trainFrameNodes.getNumNodes());
+        byte[] checkpointBytes = null;
+        if (model._parms.hasCheckpoint()) {
+            checkpointBytes = model.model_info()._boosterBytes;
+        }
+        DataInfo dataInfo = model.model_info().dataInfo();
+        BoosterParms boosterParms = XGBoostModel.createParams(model._parms, model._output.nclasses(), dataInfo.coefNames());
+        model._output._native_parameters = boosterParms.toTwoDimTable();
+        MatrixLoader loader = new FrameMatrixLoader(model, train);
+        setupTask = new XGBoostSetupTask(
+            modelKey, model._parms._save_matrix_directory, boosterParms, checkpointBytes, 
+            getRabitEnv(), trainFrameNodes._nodes, loader
+        );
+        String featureMap = XGBoostUtils.createFeatureMap(model, train);
+        featureMapFile = createFeatureMapFile(featureMap);
+    }
+    
+    @Override
+    public byte[] setup() {
+        startRabitTracker();
+        setupTask.run();
+        updateTask = new XGBoostUpdateTask(setupTask, 0).run();
+        return updateTask.getBoosterBytes();
+    }
+
+    // Don't start the tracker for 1 node clouds -> the GPU plugin fails in such a case
+    private void startRabitTracker() {
+        if (H2O.CLOUD.size() > 1) {
+            rt.start(0);
+        }
+    }
+
+    private void stopRabitTracker() {
+        if(H2O.CLOUD.size() > 1) {
+            rt.waitFor(0);
+            rt.stop();
+        }
+    }
+
+    // XGBoost seems to manipulate its frames in case of a 1 node distributed version in a way the GPU plugin can't handle
+    // Therefore don't use RabitTracker envs for 1 node
+    private Map<String, String> getRabitEnv() {
+        if(H2O.CLOUD.size() > 1) {
+            return rt.getWorkerEnvs();
+        } else {
+            return new HashMap<>();
+        }
+    }
+
+    private File createFeatureMapFile(String featureMap) {
+        try {
+            File fmFile = Files.createTempFile("h2o_xgb_" + modelKey.toString(), ".txt").toFile();
+            fmFile.deleteOnExit();
```

Comment:
Using temp files deleted on exit is a nice strategy to avoid the worst kind of leaks (restarting H2O will clean the FS), but as this file is handed over to somewhere else, having a contract that enforces it's deletion just in time would be nice. H2O needs this only for a short period of time. Just thinking out loud. Maybe some lifetime-binding mechanism, but that really feels like too much ...

Reply:
the file is deleted in the executor close method. So in theory this could be removed. will improve this a bit